File: leetcode/go/_011/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(maxArea([]int{1, 8, 6, 2, 5, 4, 8, 3, 7}))
}

func maxArea(height []int) int {
	l, r := 0, len(height)-1
	result := 0

	for l < r {
		result = max(result, (r-l)*min(height[l], height[r]))

		if height[l] > height[r] {
			r--
		} else {
			l++
		}
	}

	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

func min(a, b int) int {
	if a > b {
		return b
	}

	return a
}

File: leetcode/python/_011/main.py
from typing import List

class Solution:
    def maxArea(self, height: List[int]) -> int:
        res = 0
        l, r = 0, len(height) - 1
        while l < r:
            res = max(res, (r-l) * min(height[l], height[r]))
            if height[l] <= height[r]: l += 1
            else: r -= 1
        return res

print(Solution().maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]))